### PR TITLE
initial bits for fs mounting

### PR DESCRIPTION
**Pull request by hogsy**

### CODE REVIEW

File: platform/platform_filesystem.c [hunk 1]
```diff
@@ -52,11 +53,90 @@ For more information, please refer to <http://unlicense.org>
...
+}
+
+/**
+ * Mount the given location. On failure returns -1.
+ */
+PLFSLocation plMountLocation(const char* path) {
+    for(unsigned int i = 0; i < FS_MAX_MOUNT_LOCATIONS; ++i) {
+        FSMount* ptr = &fs_paths[i];
+        if(ptr->type == FS_MOUNT_FREE) {
+            if(plPathExists(path)) { /* attempt to mount it as a path */
```
Comment (solemnwarning) [inline platform/platform_filesystem.c]: Suggest a comment noting that `plPathExists()` only returns true for directories.

File: platform/platform_filesystem.c [hunk 2]
```diff
@@ -52,11 +53,90 @@ For more information, please refer to <http://unlicense.org>
...
+        }
+    }
+
+    ReportError(0, "failed to mount location, %s", path);
+    return -1;
+}
+
 PLresult plInitFileSystem(void) {
+    plClearMountedLocations();
+    plMountLocation(plGetWorkingDirectory()); /* todo: keep? maybe some people don't want this behaviour... */
```
Comment (solemnwarning) [inline platform/platform_filesystem.c]: I would suggest removing. People can do it themselves if they want the working directory in their search path.

File: platform/platform_filesystem.c [hunk 3]
```diff
@@ -52,11 +53,90 @@ For more information, please refer to <http://unlicense.org>
...
+ * Mount the given location. On failure returns -1.
+ */
+PLFSLocation plMountLocation(const char* path) {
+    for(unsigned int i = 0; i < FS_MAX_MOUNT_LOCATIONS; ++i) {
+        FSMount* ptr = &fs_paths[i];
+        if(ptr->type == FS_MOUNT_FREE) {
+            if(plPathExists(path)) { /* attempt to mount it as a path */
+                ptr->type = FS_MOUNT_DIR;
+                snprintf(ptr->path, sizeof(ptr->path), "%s", path);
+                fs_mount_ceiling = i;
```
Comment (solemnwarning) [inline platform/platform_filesystem.c]: Should only assign if i is greater than fs_mount_ceiling.

File: platform/platform_filesystem.c [hunk 4]
```diff
@@ -52,11 +53,90 @@ For more information, please refer to <http://unlicense.org>
...
+                return i;
+            } else { /* attempt to mount it as a package */
+                PLPackage* pkg = plLoadPackage(path);
+                if(pkg != NULL) {
+                    ptr->type = FS_MOUNT_PACKAGE;
+                    ptr->pkg = pkg;
+                    fs_mount_ceiling = i;
```
Comment (solemnwarning) [inline platform/platform_filesystem.c]: Ditto.

File: platform/platform_filesystem.c [hunk 5]
```diff
@@ -52,11 +53,90 @@ For more information, please refer to <http://unlicense.org>
 
 /*	File System	*/
 
+#define FS_MAX_MOUNT_LOCATIONS  64  /* todo: consider using a linked list instead? */
```
Comment (solemnwarning) [inline platform/platform_filesystem.c]: Sounds like a good idea to me.

**Comment by hogsy** [comment]
@solemnwarning please review latest changes and let me know if anything else needs changing.

As explained yesterday, I need to refactor the Package API so it provides a PLFile pointer when attempting to access a file (may also need read-only flag for PLFile as well? Things I'll think about for this evening.

File: platform/platform_filesystem.c [hunk 6]
```diff
@@ -53,107 +54,195 @@ For more information, please refer to <http://unlicense.org>
...
 }
 
-void plShutdownFileSystem(void) {}
+static void _plInsertMountLocation( PLFileSystemMount* location ) {
+	if ( fs_mount_root == NULL ) {
+		fs_mount_root = location;
+	}
+
+	location->prev = fs_mount_ceiling;
+	fs_mount_ceiling->next = location;
```
Comment (solemnwarning) [inline platform/platform_filesystem.c]: Should check `fs_mount_ceiling` isn't NULL before assigning into it. Also initialise `location->next` to NULL

File: platform/platform_filesystem.c [hunk 7]
```diff
@@ -293,345 +382,374 @@ void plSetWorkingDirectory(const char *path) {
...
+		if ( location->type == FS_MOUNT_DIR ) {
+			/* todo: don't allow path to search outside of mounted path */
+			snprintf( buf, sizeof( buf ), "%s/%s", location->path, path );
+			fp = _plOpenLocalFile( buf, cache );
+		} else {
+			fp = plLoadPackageFile( location->pkg, path );
+		}
+
+		if ( fp == NULL ) {
+			continue;
```
Comment (solemnwarning) [inline platform/platform_filesystem.c]: `location = location->next;` ?

File: platform/platform_filesystem.c [hunk 8]
```diff
@@ -53,107 +54,195 @@ For more information, please refer to <http://unlicense.org>
...
+static PLFileSystemMount* fs_mount_ceiling = NULL;
+
+void plClearMountedLocation( PLFileSystemMount* location ) {
+	if ( location->type == FS_MOUNT_PACKAGE ) {
+		plDestroyPackage( location->pkg );
+		location->pkg = NULL;
+	}
+
+	location->prev->next = location->next;
+	location->next->prev = location->prev;
```
Comment (solemnwarning) [inline platform/platform_filesystem.c]: Should check prev/next are non-NULL before writing into them

**Pull request merged**